src/components/legend/draw.js

```diff
@@ -568,18 +568,32 @@ function computeLegendDimensions(gd, groups, traces) {
     else {
         opts.width = 0;
         opts.height = 0;
+        var rowHeight = 0;
+        var maxTraceHeight = 0;
+        var startX = 0;
 
         traces.each(function(d) {
+
             var legendItem = d[0],
                 traceWidth = 40 + legendItem.width,
                 traceGap = opts.tracegroupgap || 5;
 
+            if((borderwidth + startX + traceGap + traceWidth) > fullLayout.width - fullLayout.margin.r) {
+                startX = 0;
+                rowHeight = rowHeight + maxTraceHeight;
+                opts.height = opts.height + maxTraceHeight;
```

Comment:
We should set `maxTraceHeight = 0` here, so that it gets recomputed for each row.